From 0bfd03fe73fb515f11c6a08a4c588ed1b7a13451 Mon Sep 17 00:00:00 2001
From: Michael Braun <michael-dev@fami-braun.de>
Date: Mon, 5 Aug 2019 22:02:42 +0200
Subject: [PATCH 6/6] nflog: add missing symbols to avoid exports being
 stripped

---
 libnl-nf-3.sym | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

--- a/libnl-nf-3.sym
+++ b/libnl-nf-3.sym
@@ -158,11 +158,16 @@ global:
 	nfnl_log_get_queue_threshold;
 	nfnl_log_msg_alloc;
 	nfnl_log_msg_get;
+	nfnl_log_msg_get_ct;
+	nfnl_log_msg_get_ct_info;
 	nfnl_log_msg_get_family;
 	nfnl_log_msg_get_gid;
 	nfnl_log_msg_get_hook;
 	nfnl_log_msg_get_hwaddr;
+	nfnl_log_msg_get_hwheader;
+	nfnl_log_msg_get_hwlen;
 	nfnl_log_msg_get_hwproto;
+	nfnl_log_msg_get_hwtype;
 	nfnl_log_msg_get_indev;
 	nfnl_log_msg_get_mark;
 	nfnl_log_msg_get_outdev;
@@ -174,12 +179,21 @@ global:
 	nfnl_log_msg_get_seq_global;
 	nfnl_log_msg_get_timestamp;
 	nfnl_log_msg_get_uid;
+	nfnl_log_msg_get_vlan_cfi;
+	nfnl_log_msg_get_vlan_id;
+	nfnl_log_msg_get_vlan_prio;
+	nfnl_log_msg_get_vlan_tag;
 	nfnl_log_msg_put;
+	nfnl_log_msg_set_ct;
+	nfnl_log_msg_set_ct_info;
 	nfnl_log_msg_set_family;
 	nfnl_log_msg_set_gid;
 	nfnl_log_msg_set_hook;
 	nfnl_log_msg_set_hwaddr;
+	nfnl_log_msg_set_hwheader;
+	nfnl_log_msg_set_hwlen;
 	nfnl_log_msg_set_hwproto;
+	nfnl_log_msg_set_hwtype;
 	nfnl_log_msg_set_indev;
 	nfnl_log_msg_set_mark;
 	nfnl_log_msg_set_outdev;
@@ -191,13 +205,20 @@ global:
 	nfnl_log_msg_set_seq_global;
 	nfnl_log_msg_set_timestamp;
 	nfnl_log_msg_set_uid;
+	nfnl_log_msg_set_vlan_tag;
+	nfnl_log_msg_test_ct;
+	nfnl_log_msg_test_ct_info;
 	nfnl_log_msg_test_gid;
 	nfnl_log_msg_test_hook;
+	nfnl_log_msg_test_hwheader;
+	nfnl_log_msg_test_hwlen;
 	nfnl_log_msg_test_hwproto;
+	nfnl_log_msg_test_hwtype;
 	nfnl_log_msg_test_mark;
 	nfnl_log_msg_test_seq;
 	nfnl_log_msg_test_seq_global;
 	nfnl_log_msg_test_uid;
+	nfnl_log_msg_test_vlan_tag;
 	nfnl_log_pf_bind;
 	nfnl_log_pf_unbind;
 	nfnl_log_put;
